### PR TITLE
large range scan support

### PR DESCRIPTION
**Pull request by pyr**
Large range scans may happen in the record layer by either
scanning explicit ranges or execute queries for which the
plan determines a scan should be performed. While this is not
something one should optimize for, it is legitimate to want
to perform large scans on occasion.

In this case, it may be ok to allow relaxing consistency
guarantees. To achieve that, we introduce a low-level
`continuation-traversing-reducer` function, used by
`long-range-reducer` and `long-query-reducer`.

A lot of help from @arnaudgeiser, especially around fixing
prefix range building.

### CODE REVIEW

File: src/exoscale/vinyl/cursor.clj [hunk 1]
```diff
@@ -14,23 +14,30 @@
...
-                  (let [next?   (.hasNext result)
-                        new-acc (when next? (swap! acc f (.get result)))]
-                    (and (not (reduced? new-acc)) next?))))))))
-      (.getExecutor cursor))
-     (fn/make-fun (fn [_] (unreduced @acc))))))
+   gets in which will provide a way to do this directly from record layer.
+
+   When `cont-fn` is given, it will be called on the last seen continuation
+   byte array for every new element."
+  ([^RecordCursor cursor f init cont-fn]
```
Comment (pyr) [inline src/exoscale/vinyl/cursor.clj:21]: A new arity is introduced. If non nil, `cont-f` will get called with each continuation

File: src/exoscale/vinyl/store.clj [hunk 2]
```diff
@@ -45,7 +49,25 @@
     Protocolized so it can be called against the database or
     the store")
   (get-metadata [this]
-    "Return this context's record metadata"))
+    "Return this context's record metadata")
+  (new-runner [this]
+    "Return a runner to handle retryable logic"))
+
+(defn ^FDBDatabaseRunner runner-opts
```
Comment (pyr) [inline src/exoscale/vinyl/store.clj:59]: Sadly, `cond->` cannot be used here, all mutating functions on the runner return nil

File: src/exoscale/vinyl/store.clj [hunk 3]
```diff
@@ -138,6 +162,29 @@
             (apply [_ context]
               (f (store-from-builder (::builder this) context)))))))
 
+(defn wrapped-runner
```
Comment (pyr) [inline src/exoscale/vinyl/store.clj:168]: A convenience wrapper around a `FDBDatabaseRunner` to ensure it behaves like a database.

File: src/exoscale/vinyl/cursor.clj [hunk 4]
```diff
@@ -14,23 +14,30 @@
...
+       (reify Supplier
+         (get [_]
+           (-> cursor
+               .onNext
+               (.thenApply
+                (fn/make-fun
+                 (fn [^RecordCursorResult result]
+                   (when (ifn? cont-fn)
+                     (-> result .getContinuation .toBytes cont-fn))
+                   (let [next?   (.hasNext result)
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/cursor.clj:34]: Using `hasNext` is probably totally fine for our usage.
However, we won't handle correctly the case where we have a cursor defined with `limitRowsTo`.
If we want to support it, we would have to also check `hasStoppedBeforeEnd`: https://github.com/FoundationDB/fdb-record-layer/blob/e5dae5553833b50df82a67e884e1349bd6953053/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java#L222

File: src/exoscale/vinyl/store.clj [hunk 5]
```diff
@@ -45,7 +49,25 @@
...
+    "Return a runner to handle retryable logic"))
+
+(defn ^FDBDatabaseRunner runner-opts
+  [^FDBDatabaseRunner runner
+   {::keys [max-attempts
+            initial-delay
+            max-delay
+            transaction-timeout]}]
+  (when (some? max-attempts)
+    (.setMaxAttempts runner (int max-attempts)))
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj:66]: __Info__

`setMaxAttempts` is the most relevant parameter for us as we'll be able to rely on the transaction retry mechanism to exhaust a range of keys.

File: src/exoscale/vinyl/store.clj [hunk 6]
```diff
@@ -302,3 +384,52 @@
...
+   (long-query-reducer db f init query {}))
+  ([db f init query {::keys [values] :as opts}]
+   (let [cont    (atom nil)
+         result  (atom init)
+         props   (execute-properties (dissoc opts ::limit))
+         ctx     (if (some? values) (query/bindings values) query/empty-context)
+         q       (as-query query)
+         runner  (wrapped-runner db
+                                 {::max-attempts        Integer/MAX_VALUE
+                                   ::max-delay           2
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj]: As for huge buckets we'll probably use this value a lot, I would expect to see the equivalent of `try to do it as soon as possible`. Maybe we can even use 0?

BTW: The default value is 1000ms which is way too high for what we are doing. 2ms is probably acceptable.

File: src/exoscale/vinyl/store.clj [hunk 7]
```diff
@@ -302,3 +384,52 @@
...
+                                  ::initial-delay       2})]
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj]: __Info__

The default value is 10ms.
We should always match both `max-delay` and `initial-delay`.

File: src/exoscale/vinyl/store.clj [hunk 8]
```diff
@@ -45,7 +49,25 @@
...
+            max-delay
+            transaction-timeout]}]
+  (when (some? max-attempts)
+    (.setMaxAttempts runner (int max-attempts)))
+  (when (some? initial-delay)
+    (.setInitialDelayMillis runner (long initial-delay)))
+  (when (some? max-delay)
+    (.setMaxDelayMillis runner (long max-delay)))
+  (when (some? transaction-timeout)
+    (.setTransactionTimeoutMillis runner (long transaction-timeout)))
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj:72]: __Info__

This transaction timeout is not related to the five seconds transaction.
It's actually a non-retriable Exception: https://github.com/FoundationDB/fdb-record-layer/blame/036443fe91fe822422fe0d732bfade3c91d68441/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunner.java#L193-L194

File: src/exoscale/vinyl/store.clj (hunk 3 above)
Comment (arnaudgeiser) [reply src/exoscale/vinyl/store.clj:168]: So smart, with that we are finally able to have complete control over the runner instead of passing the `FDBDatabase`

File: src/exoscale/vinyl/store.clj [hunk 9]
```diff
@@ -275,13 +328,42 @@
...
+            skip
+            limit]
+    :as    props}]
+  (if (nil? props)
+    ExecuteProperties/SERIAL_EXECUTE
+    (-> (ExecuteProperties/newBuilder)
+        (.setFailOnScanLimitReached (boolean fail-on-scan-limit-reached?))
+        (.setIsolationLevel (as-isolation-level isolation-level))
+        (cond-> (some? skip) (.setSkip (int skip)))
+        (cond-> (some? limit) (.setReturnedRowLimit (int limit)))
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj:400]: As mentioned above, I don't think this one will work anymore.
This was probably the only thing `AutoContinuingCursor` was handling correctly.

File: src/exoscale/vinyl/store.clj [hunk 10]
```diff
@@ -302,3 +384,52 @@
...
+  ;; To that effect, a runner is created with specific parameters (a large
+  ;; `max-attempts` value, as well as minimum viable delays.
+  ;;
+  ;; We then call `apply-reduce` on the returned cursor from our query with a
+  ;; twist: every visited element will get its continuation stored in an atom.
+  ;; When interrupted, the function will be retried, which pops the last seen
+  ;; continuation.
+  ;;
+  ([db f init query]
+   (long-query-reducer db f init query {}))
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj]: In apply-transforms we use the keys `reducer` and `reduce-init`.
Maybe we could use the same names (at least for `f`) ?

https://github.com/exoscale/vinyl/blob/master/src/exoscale/vinyl/cursor.clj#L38-L39

File: src/exoscale/vinyl/store.clj [hunk 11]
```diff
@@ -302,3 +384,52 @@
...
+                                  ::initial-delay       2})]
+     (-> (run-async
+          runner
+          (fn [^FDBRecordStore store]
+            (-> (.execute (.planQuery store q)
+                          store
+                          ctx
+                          ^bytes @cont
+                          props)
+                (cursor/apply-reduce f result #(reset! cont %)))))
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/store.clj]: __Info__

This is where the magic happens to put the continuation into the atom to be able to restart the transaction at the last reading key.

**Review by arnaudgeiser — APPROVED** [review]
Some comments.
I would also see some changes on `prefix-range` and `scan-range` to use this mechanism, but it can be part of another PR.

File: src/exoscale/vinyl/fn.clj [hunk 12]
```diff
@@ -2,12 +2,38 @@
...
            java.util.function.BiFunction
-           java.util.function.Consumer))
+           java.util.function.Consumer
+           java.util.function.BiConsumer
+           java.util.function.Supplier
+           java.util.concurrent.CompletableFuture
+           java.lang.AutoCloseable))
 
 (defn ^Function make-fun
   "Turn a function of one argument into a java Function"
```
Comment (arnaudgeiser) [inline src/exoscale/vinyl/fn.clj]: The description is not totally accurate here.
To be able to use this function correctly, the parameter `f` needs to have 1-arity AND 2-arities.
Otherwise, you will be able to only use it as a `BiConsumer/BiFunction` OR as a `Function/Consumer`.

File: src/exoscale/vinyl/store.clj (hunk 11 above)
Comment (mping-exo) [reply src/exoscale/vinyl/store.clj]: The atom is just for "convenience" right? Due to the dependency between result pages (needs the cont token for the next page) the code will run effectively serial, even if it's scheduled on some thread right?

Comment (arnaudgeiser) [reply src/exoscale/vinyl/store.clj]: Yes, we may use volatile! instead.

**Review by arnaudgeiser — APPROVED** [review]

File: src/exoscale/vinyl/store.clj (hunk 11 above)
Comment (mping-exo) [reply src/exoscale/vinyl/store.clj]: I'm fine with atom, I was just unsure.

File: src/exoscale/vinyl/fn.clj (hunk 12 above)
Comment (pyr) [reply src/exoscale/vinyl/fn.clj]: Thanks, I used a small trick with `definterface` to make things a bit cleaner

File: src/exoscale/vinyl/cursor.clj (hunk 4 above)
Comment (pyr) [reply src/exoscale/vinyl/cursor.clj:34]: Keeping with the current approach, it makes sense to keep things as they are for now.

File: src/exoscale/vinyl/store.clj (hunk 8 above)
Comment (pyr) [reply src/exoscale/vinyl/store.clj:72]: Agreed. Leaving things here since they may be used elsewhere / later on

File: src/exoscale/vinyl/store.clj (hunk 9 above)
Comment (pyr) [reply src/exoscale/vinyl/store.clj:400]: Agreed, but scan-props can be fed to other functions than the large scanning functions.

File: src/exoscale/vinyl/store.clj (hunk 10 above)
Comment (pyr) [reply src/exoscale/vinyl/store.clj]: I wanted to match the args to `reduce` (but seems I missed the target since they have `val`, not `init`). I will adapt to match `reduce` completely

**Review by arnaudgeiser — APPROVED** [review]
Final words, that's good enough for being shipped.
However, I wonder if it would be better to extend `long-range-reducer` where the last entry of `items` can optionally be a vector containing a `prefix` and a `marker`.

```clojure
(store/long-range-reducer db f val record-type [bucket [prefix marker]])
```

Example:
```clojure
;; with marker
(store/long-range-reducer *db* incrementor 0 :Object ["small-bucket" ["files/" "files/00000010.txt"]])

;; without marker
(store/long-range-reducer *db* incrementor 0 :Object ["small-bucket" "files/"])
```
But we need to move on with what we have currently.
